data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/src/main/java/org/eclipse/edc/protocol/dsp/negotiation/api/controller/DspNegotiationController.java

```diff
@@ -0,0 +1,282 @@
+
+    private final IdentityService identityService;
+    private final JsonLdTransformerRegistry transformerRegistry;
+    private final String callbackAddress;
+    private final Monitor monitor;
+    private final ContractNegotiationService service;
+    private final ContractNegotiationProtocolService protocolService;
+    private final ObjectMapper mapper;
+
+    public DspNegotiationController(Monitor monitor, TypeManager typeManager, String callbackAddress,
+                                    IdentityService identityService, JsonLdTransformerRegistry transformerRegistry,
+                                    ContractNegotiationService service, ContractNegotiationProtocolService protocolService) {
+        this.callbackAddress = callbackAddress;
+        this.identityService = identityService;
+        this.monitor = monitor;
+        this.protocolService = protocolService;
+        this.service = service;
+        this.transformerRegistry = transformerRegistry;
+
+        this.mapper = typeManager.getMapper(TYPE_MANAGER_CONTEXT_JSON_LD);
+    }
+
+    @GET
+    @Path("{id}")
+    @Override
+    public Map<String, Object> getNegotiation(@PathParam("id") String id, @HeaderParam(HttpHeaders.AUTHORIZATION) String token) {
+        monitor.debug(format("DSP: Incoming request for contract negotiation with id %s", id));
+
+        checkAndReturnAuthToken(token);
+
+        var negotiation = service.findbyId(id);
+        if (negotiation == null) {
+            throw new ObjectNotFoundException(ContractNegotiation.class, id);
+        }
+
+        var result = transformerRegistry.transform(negotiation, JsonObject.class).orElseThrow((failure ->
+                new EdcException(format("Failed to build response: %s", failure.getFailureDetail()))));
+
+        return mapper.convertValue(compact(result, jsonLdContext()), Map.class);
+    }
+
+    @POST
+    @Path(INITIAL_CONTRACT_REQUEST)
+    @Override
+    public Map<String, Object> initiateNegotiation(@RequestBody(description = "dspace:ContractRequestMessage", required = true) JsonObject body, @HeaderParam(HttpHeaders.AUTHORIZATION) String token) {
+        monitor.debug("DSP: Incoming ContractRequestMessage for initiating a contract negotiation.");
+
+        var claimToken = checkAndReturnAuthToken(token);
+
+        var expanded = expand(body).getJsonObject(0);
+        var message = transformerRegistry.transform(hasValidType(expanded, DSPACE_NEGOTIATION_CONTRACT_REQUEST_MESSAGE), ContractRequestMessage.class)
+                .orElseThrow(failure -> new InvalidRequestException(format("Request body was malformed: %s", failure.getFailureDetail())));
+
+        var negotiation = protocolService.notifyRequested(message, claimToken).orElseThrow(exceptionMapper(ContractNegotiation.class));
+
+        var result = transformerRegistry.transform(negotiation, JsonObject.class).orElseThrow(failure ->
+                new EdcException(format("Failed to build response: %s", failure.getFailureDetail())));
```

Comment:
In case of failure shouldn't a `NegotiationError` message returned? could be something that an `ExceptionMapper` could take care of?

Reply:
Not important for phase one. We need to think of a nice way to handle those for all three processes: catalog, negotiation, transfer. See @ronjaquensel comment in #2784. I created #2796.

Reply:
This type of error (transform) should be mapped to HTTP 400. The spec also mandates JSON-LD as the content type. Why not introduce an exception mapper or create the response via a generic method that can be reused?

Reply:
Transformer errors are mapped to 400. Only response creation fails are mapped to 500 and this is correct. Adding DSP error handling will be done in a different PR.